### PR TITLE
Feature: add sns support

### LINKED ISSUE #10: add support for SNS

### PR DESCRIPTION
**Pull request by erezrokah**
Fixes #10

* This is mostly a reproduction of the process described [here](https://www.alexdebrie.com/posts/aws-api-gateway-service-proxy/#creating-the-service-proxy-integration).
* I removed some [unnecessary](https://daniel.haxx.se/blog/2015/09/11/unnecessary-use-of-curl-x/) `-X POST`  from the readme.
* Integrations tests will validate the content of the published message by attaching an SQS queue to the SNS topic (I'm using https://github.com/erezrokah/aws-testing-library, so if you don't want to introduce another dependency I can remove the content validation or implement the functionally to validate it as a part of this plugin - also I don't want to feel like I'm promoting my lib for no reason).
* Added a specific integration test to validate the custom mapping template
* Fixed some copy paste errors in `lib/package/kinesis/compileMethodsToKinesis.test.js`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -307,4 +335,31 @@ custom:
...
+custom:
+  apiGatewayServiceProxies:
+    - kinesis:
+        path: /sns
+        method: post
+        topicName: { 'Fn::GetAtt': ['SNSTopic', 'TopicName'] }
+        request:
+          template:
+            application/json:
+              'Fn::Join':
```
Comment (theburningmonk) [inline README.md]: can you change this to use `Fn::Sub` instead? e.g.

```
'Fn::Sub': "Action=Publish&Message=$util.urlEncode('This is a fixed message')&TopicArn=$util.urlEncode('${MyTopic}')"
```

Comment (erezrokah) [reply README.md]: Of course

Comment (theburningmonk) [reply README.md]: good point, then maybe mention the `serverless-pseudo-parameters` as well? `Fn::Join` is just so unwildy and hard to follow

File: __tests__/integration/sns/custom-mapping-template/tests.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+describe('Single SNS Proxy Integration Test', () => {
+  let endpoint
+  let stage
+  let region
+  let queueUrl
+  const config = '__tests__/integration/sns/custom-mapping-template/service/serverless.yml'
+
+  beforeAll(async () => {
+    ;({
```
Comment (theburningmonk) [inline __tests__/integration/sns/custom-mapping-template/tests.js]: why the leading `;` here?

Comment (theburningmonk) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: is this a js thing? looks funky lol

File: __tests__/integration/sns/custom-mapping-template/tests.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    expect(json.PublishResponse.ResponseMetadata).to.have.own.property('RequestId')
+
+    const expectedFixedMessage = 'This is a fixed message'
+    await expect({
+      region,
+      queueUrl,
+      timeout: 10000,
+      pollEvery: 2500
+    }).to.have.message((message) => message.Message === expectedFixedMessage)
```
Comment (theburningmonk) [inline __tests__/integration/sns/custom-mapping-template/tests.js]: I see, this is what the `aws-testing-library` gives you, yeah, this is nice. Good use case :-)

File: __tests__/integration/sns/custom-mapping-template/tests.js (hunk 2 above)
Comment (erezrokah) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: It's a js thing since the auto formatting for the repo doesn't have trailing `;`.
I would rather just add `;` at the end of each line

Comment (theburningmonk) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: ok, what's wrong with this?

```
beforeAll(async () => {
  { stage, region, endpoint, outputs: { SqsQueueUrl: queueUrl } } = await deployWithRandomStage(config)
})
```
Does the code you have there behalve differently?

File: README.md (hunk 1 above)
Comment (erezrokah) [reply README.md]: Updated the readme. I referenced`serverless-cloudformation-sub-variables`.
`serverless-pseudo-parameters` lets you do the same?

File: __tests__/integration/sns/custom-mapping-template/tests.js (hunk 2 above)
Comment (erezrokah) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: That's actually the main reason I didn't use that method before, but when you have many variables you need to destruct it is still nicer that multiple lines

File: README.md (hunk 1 above)
Comment (theburningmonk) [reply README.md]: yup, the two kinda do the same thing

File: __tests__/integration/sns/custom-mapping-template/tests.js (hunk 2 above)
Comment (erezrokah) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: You can't destruct into existing variables without wrapping the line with `()`

Comment (theburningmonk) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: Sorry, I'm confused. So the snippet I had does the same thing? If that's the case, why the leading `;`?

Comment (theburningmonk) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: ah, do you mean, have them on separate lines, like this?

```
beforeAll(async () => {
  { stage, 
    region, 
    endpoint, 
    outputs: { SqsQueueUrl: queueUrl } 
  } = await deployWithRandomStage(config)
})
```
is this not valid?

Comment (erezrokah) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: We could just change `prettier.config.js` to use `semi: true`, but we can do it in another maintenance pr.

Comment (theburningmonk) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: ah! that's right, thanks for reminding me ;-)

> You can't destruct into existing variables without wrapping the line with `()`.
> You get a syntax error

and if you have `()` wrapping the whole thing then you also get a syntax error without the leading `;`.. right I think I get it now

Comment (erezrokah) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: Sorry I'll try to be more clear.

You can do:

```
let { a } = result
```

but not

```
let a

{ a } = result
```

You can do:

```
let a;

({ a } = result);
```

or
 
```
let a
;({ a } = result))
```
or

```
let a 

a = result.a
```

File: README.md (hunk 1 above)
Comment (erezrokah) [reply README.md]: Should I change it to `serverless-pseudo-parameters`?

Comment (theburningmonk) [reply README.md]: nah, it's fine, leave it, they do the same thing

File: __tests__/integration/sns/custom-mapping-template/tests.js (hunk 2 above)
Comment (horike37) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: Ok, we can discuss about the config in another issue. Actually, either one is fine for me.

Comment (theburningmonk) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: cheers, thanks for the detailed explanation 👍

File: lib/package/sns/compileMethodsToSns.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  getSnsMethodIntegration(http) {
+    const integration = {
+      IntegrationHttpMethod: 'POST',
+      Type: 'AWS',
+      Credentials: {
+        'Fn::GetAtt': ['ApigatewayToSnsRole', 'Arn']
+      },
+      Uri: {
+        'Fn::Join': [
```
Comment (theburningmonk) [inline lib/package/sns/compileMethodsToSns.js]: how about `Fn::Sub` here instead? i.e. `Fn::Sub`: 'arn:aws:apigateway:${AWS::Region}:sns:path//'`

File: README.md [hunk 5]
```diff
@@ -102,7 +103,7 @@ resources:
 Sample request after deploying.
 
 ```bash
-curl -X POST https://xxxxxx.execute-api.us-east-1.amazonaws.com/dev/sqs -d '{"message": "testtest"}' -H 'Content-Type:application/json'
+curl https://xxxxxx.execute-api.us-east-1.amazonaws.com/dev/sqs -d '{"message": "testtest"}' -H 'Content-Type:application/json'
```
Comment (horike37) [inline README.md]: Why removing? `curl` default behavior is GET, right?
Those examples are specified `post` in `method` param.

File: lib/package/sns/compileMethodsToSns.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+  buildDefaultSnsRequestTemplate(http) {
+    const { topicName } = http
+
+    const topicArn = {
+      'Fn::Sub': ['arn:aws:sns:${AWS::Region}:${AWS::AccountId}:${topicName}', { topicName }]
+    }
+
+    return {
+      'Fn::Join': [
```
Comment (theburningmonk) [inline lib/package/sns/compileMethodsToSns.js]: same here, can we use `Fn::Sub` instead?

File: lib/package/sns/validateSnsServiceProxy.test.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+      ]
+    }
+
+    expect(() => serverlessApigatewayServiceProxy.validateSnsServiceProxy()).to.throw(
+      serverless.classes.Error
+    )
+  })
+
+  it('should validate the "topicName" property if it is string or AWS intrinsic function', async () => {
+    serverlessApigatewayServiceProxy.validated = {
```
Comment (theburningmonk) [inline lib/package/sns/validateSnsServiceProxy.test.js]: I'd argue the test doesn't match the test name here - the tests checks that we throw when `topicName` is an array and does nothing to check whether it's string or intrinsic function.

File: lib/package/sns/validateSnsServiceProxy.test.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        }
+      ]
+    }
+
+    expect(() => serverlessApigatewayServiceProxy.validateSnsServiceProxy()).to.throw(
+      serverless.classes.Error
+    )
+  })
+
+  it('should validate the "topicName" property that AWS intrinsic function is correct syntax', async () => {
```
Comment (theburningmonk) [inline lib/package/sns/validateSnsServiceProxy.test.js]: similarly, here, the test is really 'it should throw if topicName is an intrinsic function that's not Fn::GetAtt'

File: lib/package/sns/validateSnsServiceProxy.test.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        }
+      ]
+    }
+
+    expect(() => serverlessApigatewayServiceProxy.validateSnsServiceProxy()).to.throw(
+      serverless.classes.Error
+    )
+  })
+
+  it('should validate the "topicName" property if wrong AWS intrinsic function is specified', async () => {
```
Comment (theburningmonk) [inline lib/package/sns/validateSnsServiceProxy.test.js]: sorry to keep barking on the whole test naming thing, but the test here is that we throw when the Fn::GetAtt intrinsic function asks for a property other than TopicName

File: lib/package/sns/compileMethodsToSns.js (hunk 4 above)
Comment (erezrokah) [reply lib/package/sns/compileMethodsToSns.js]: sure I'll update it

File: README.md (hunk 5 above)
Comment (erezrokah) [reply README.md]: I found out about it recently:
https://daniel.haxx.se/blog/2015/09/11/unnecessary-use-of-curl-x/

If you specify `-d` curl will send a `POST request.

`-X` should be used when you actually want to override the method

File: lib/package/sns/compileMethodsToSns.js (hunk 6 above)
Comment (erezrokah) [reply lib/package/sns/compileMethodsToSns.js]: Sure

File: lib/package/sns/validateSnsServiceProxy.test.js (hunk 7 above)
Comment (erezrokah) [reply lib/package/sns/validateSnsServiceProxy.test.js]: I agree. I'll think of a better name :)

File: lib/package/sns/validateSnsServiceProxy.test.js (hunk 8 above)
Comment (erezrokah) [reply lib/package/sns/validateSnsServiceProxy.test.js]: I agree again :)

File: lib/package/sns/validateSnsServiceProxy.test.js (hunk 9 above)
Comment (erezrokah) [reply lib/package/sns/validateSnsServiceProxy.test.js]: Correct, I'll update it

**Comment by erezrokah** [comment]
Thank you for the comments, they make a lot of sense.
Please let me know if there is something more :)

File: README.md (hunk 5 above)
Comment (horike37) [reply README.md]: ok, I didn't know that. thank you for letting me know 😄

Comment (erezrokah) [reply README.md]: I've read that post several times, been using `curl` without `-X` for a few weeks now and still can't get my head around it :smiley:

**Comment by theburningmonk** [comment]
@erezrokah no worries, I learnt something new myself too, so thank you!

**Comment by theburningmonk** [comment]
LGTM! @horike37 ?

File: __tests__/integration/sns/custom-mapping-template/tests.js (hunk 3 above)
Comment (horike37) [reply __tests__/integration/sns/custom-mapping-template/tests.js]: Love it.
